### PR TITLE
Array functions concat slice push pop

### PR DESCRIPTION
**Pull request by KochetovNicolai**

### CODE REVIEW

**Comment by KochetovNicolai** [comment]
autotest

File: dbms/src/Functions/FunctionsArray.cpp [hunk 1]
```diff
@@ -31,13 +32,13 @@ String FunctionArray::getName() const
     return name;
 }
 
-FunctionPtr FunctionArray::create(const Context & context)
+FunctionPtr FunctionArray::create(const Context &context)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsArray.cpp]: .

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/src/Common/TypeList.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply dbms/src/Common/TypeList.h]: unified

File: dbms/src/Common/TypeList.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+{
+};
+
+/// Append Type to TypeList
+/// Usage:
+///     using TypeListWithType = typename AppendToTypeList<Type, ConcreteTypeList>::Type;
+template <typename TypeToAppend, typename List, typename ... Types>
+struct AppendToTypeList
+{
+    using Type = typename AppendToTypeList<TypeToAppend, typename List::Tail, Types ..., typename List::Head>::Type;
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply dbms/src/Common/TypeList.h]: TypeList doesn't get permuted. New type is pushed as head.
Renamed to PrependToTypeList

File: dbms/src/Functions/GatherUtils.h [hunk 4]
```diff
@@ -728,24 +891,70 @@ inline ALWAYS_INLINE void writeSlice(const StringSource::Slice & slice, FixedStr
...
 {
-    sink.elements.insertRangeFrom(*slice.elements, slice.begin, slice.size);
+    if (typeid(slice.elements) == typeid(&sink.elements))
+        sink.elements.insertRangeFrom(*slice.elements, slice.begin, slice.size);
+    else
+    {
+        for (size_t i = 0; i < slice.size; ++i)
+        {
+            Field field;
+            slice.elements->get(slice.begin + i, field);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply dbms/src/Functions/GatherUtils.h]: There is a comment above that columns should have the same type.

**Pull request merged**